refactor(isolated_declarations): do not store temp values in arena

overlookmotel opened the pull request:
`stmts` here is an `ArenaVec<&Statement>` (`&Statement` not `Statement`). This is a temporary collection, and doesn't end up in the AST, so we shouldn't store it in the arena.

Use a `std::vec::Vec` instead of `oxc_allocator::Vec`.

This should also remove some of the lifetime problems we have in #9656.

This change may cause a small perf hit, but in my view it's still a good change. If we want to get a speed boost, a better solution would be to have a temporary "scratch space" arena which we can allocate *all* temporary values into. This would likely give us a sizeable speed boost across many parts of Oxc (https://github.com/oxc-project/backlog/issues/121).

overlookmotel commented:
* **#9733** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9733?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9733?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9733).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing reviewed (approved):
The changes look good to me!

Dunqing commented:
### Merge activity

* **Mar 12, 11:09 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Mar 12, 11:09 PM EDT**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
refactor(isolated_declarations): do not store temp values in arena (#9733)

`stmts` here is an `ArenaVec<&Statement>` (`&Statement` not `Statement`). This is a temporary collection, and doesn't end up in the AST, so we shouldn't store it in the arena.

Use a `std::vec::Vec` instead of `oxc_allocator::Vec`.

This should also remove some of the lifetime problems we have in #9656.

This change may cause a small perf hit, but in my view it's still a good change. If we want to get a speed boost, a better solution would be to have a temporary "scratch space" arena which we can allocate *all* temporary values into. This would likely give us a sizeable speed boost across many parts of Oxc (https://github.com/oxc-project/backlog/issues/121).

overlookmotel commented:
I've taken off the merge label in hope of stopping Graphite - but maybe it's unstoppable! Would like to see the benchmarks first ideally.

Dunqing commented:
> I've taken off the merge label in hope of stopping Graphite - but maybe it's unstoppable! Would like to see the benchmarks first ideally.

Only 0.2ms dropped

<img width="754" alt="image" src="https://github.com/user-attachments/assets/c8b05525-3b4e-4ec8-8c76-f90542de19e8" />

overlookmotel commented:
-0.3% perf hit. That's no so bad.
